From 6e41fa2e5cd84d0970f1ee1d9c9767dbc974952a Mon Sep 17 00:00:00 2001
From: Michael Meskes <meskes@postgresql.org>
Date: Fri, 5 Oct 2012 16:37:45 +0200
Subject: [PATCH] Fixed test for array boundary.

Instead of continuing if the next character is not an array boundary get_data()
used to continue only on finding a boundary so it was not able to read any
element after the first.

Bug-Ubuntu: https://launchpad.net/bugs/1063613
---
 src/interfaces/ecpg/ecpglib/data.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/src/interfaces/ecpg/ecpglib/data.c b/src/interfaces/ecpg/ecpglib/data.c
index db97503..dcccd92 100644
--- a/src/interfaces/ecpg/ecpglib/data.c
+++ b/src/interfaces/ecpg/ecpglib/data.c
@@ -422,6 +422,7 @@ ecpg_get_data(const PGresult *results, int act_tuple, int act_field, int lineno,
 							ecpg_raise(lineno, ECPG_CONVERT_BOOL,
 									   ECPG_SQLSTATE_DATATYPE_MISMATCH,
 									   NULL);
+						pval++;
 						break;
 					}
 					else if (pval[0] == 't' && pval[1] == '\0')
@@ -434,6 +435,7 @@ ecpg_get_data(const PGresult *results, int act_tuple, int act_field, int lineno,
 							ecpg_raise(lineno, ECPG_CONVERT_BOOL,
 									   ECPG_SQLSTATE_DATATYPE_MISMATCH,
 									   NULL);
+						pval++;
 						break;
 					}
 					else if (pval[0] == '\0' && PQgetisnull(results, act_tuple, act_field))
@@ -777,7 +779,7 @@ ecpg_get_data(const PGresult *results, int act_tuple, int act_field, int lineno,
 					++pval;
 			}
 		}
-	} while (*pval != '\0' && array_boundary(isarray, *pval));
+	} while (*pval != '\0' && !array_boundary(isarray, *pval));
 
 	return (true);
 }
-- 
1.7.2.5

